File: javascript/src/api_internal/streamStream.ts
// this file is @generated
import {
  ListResponseStreamOut,
  ListResponseStreamOutSerializer,
} from "../models/listResponseStreamOut";
import { Ordering } from "../models/ordering";
import { StreamIn, StreamInSerializer } from "../models/streamIn";
import { StreamOut, StreamOutSerializer } from "../models/streamOut";
import { StreamPatch, StreamPatchSerializer } from "../models/streamPatch";
import { HttpMethod, SvixRequest, SvixRequestContext } from "../request";

export interface StreamStreamListOptions {
  /** Limit the number of returned items */
  limit?: number;
  /** The iterator returned from a prior invocation */
  iterator?: string | null;
  /** The sorting order of the returned items */
  order?: Ordering;
}

export interface StreamStreamCreateOptions {
  idempotencyKey?: string;
}

export class StreamStream {
  public constructor(private readonly requestCtx: SvixRequestContext) {}

  /** List of all the organization's streams. */
  public list(options?: StreamStreamListOptions): Promise<ListResponseStreamOut> {
    const request = new SvixRequest(HttpMethod.GET, "/api/v1/stream");

    request.setQueryParam("limit", options?.limit);
    request.setQueryParam("iterator", options?.iterator);
    request.setQueryParam("order", options?.order);

    return request.send(this.requestCtx, ListResponseStreamOutSerializer._fromJsonObject);
  }

  /** Creates a new stream. */
  public create(
    streamIn: StreamIn,
    options?: StreamStreamCreateOptions
  ): Promise<StreamOut> {
    const request = new SvixRequest(HttpMethod.POST, "/api/v1/stream");

    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.setBody(StreamInSerializer._toJsonObject(streamIn));

    return request.send(this.requestCtx, StreamOutSerializer._fromJsonObject);
  }

  /** Get a stream by id or uid. */
  public get(streamId: string): Promise<StreamOut> {
    const request = new SvixRequest(HttpMethod.GET, "/api/v1/stream/{stream_id}");

    request.setPathParam("stream_id", streamId);

    return request.send(this.requestCtx, StreamOutSerializer._fromJsonObject);
  }

  /** Update a stream. */
  public update(streamId: string, streamIn: StreamIn): Promise<StreamOut> {
    const request = new SvixRequest(HttpMethod.PUT, "/api/v1/stream/{stream_id}");

    request.setPathParam("stream_id", streamId);
    request.setBody(StreamInSerializer._toJsonObject(streamIn));

    return request.send(this.requestCtx, StreamOutSerializer._fromJsonObject);
  }

  /** Delete a stream. */
  public delete(streamId: string): Promise<void> {
    const request = new SvixRequest(HttpMethod.DELETE, "/api/v1/stream/{stream_id}");

    request.setPathParam("stream_id", streamId);

    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Partially update a stream. */
  public patch(streamId: string, streamPatch: StreamPatch): Promise<StreamOut> {
    const request = new SvixRequest(HttpMethod.PATCH, "/api/v1/stream/{stream_id}");

    request.setPathParam("stream_id", streamId);
    request.setBody(StreamPatchSerializer._toJsonObject(streamPatch));

    return request.send(this.requestCtx, StreamOutSerializer._fromJsonObject);
  }
}

File: svix-cli/src/cmds/api/stream_stream.rs
// this file is @generated
use clap::{Args, Subcommand};
use svix::api::*;

#[derive(Args, Clone)]
pub struct StreamStreamListOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// The sorting order of the returned items
    #[arg(long)]
    pub order: Option<Ordering>,
}

impl From<StreamStreamListOptions> for svix::api::StreamStreamListOptions {
    fn from(value: StreamStreamListOptions) -> Self {
        let StreamStreamListOptions {
            limit,
            iterator,
            order,
        } = value;
        Self {
            limit,
            iterator,
            order,
        }
    }
}

#[derive(Args, Clone)]
pub struct StreamStreamCreateOptions {
    #[arg(long)]
    pub idempotency_key: Option<String>,
}

impl From<StreamStreamCreateOptions> for svix::api::StreamStreamCreateOptions {
    fn from(value: StreamStreamCreateOptions) -> Self {
        let StreamStreamCreateOptions { idempotency_key } = value;
        Self { idempotency_key }
    }
}

#[derive(Args)]
#[command(args_conflicts_with_subcommands = true, flatten_help = true)]
pub struct StreamStreamArgs {
    #[command(subcommand)]
    pub command: StreamStreamCommands,
}

#[derive(Subcommand)]
pub enum StreamStreamCommands {
    /// List of all the organization's streams.
    List {
        #[clap(flatten)]
        options: StreamStreamListOptions,
    },
    /// Creates a new stream.
    Create {
        stream_in: crate::json::JsonOf<StreamIn>,
        #[clap(flatten)]
        options: StreamStreamCreateOptions,
    },
    /// Get a stream by id or uid.
    Get { stream_id: String },
    /// Update a stream.
    Update {
        stream_id: String,
        stream_in: crate::json::JsonOf<StreamIn>,
    },
    /// Delete a stream.
    Delete { stream_id: String },
    /// Partially update a stream.
    Patch {
        stream_id: String,
        stream_patch: Option<crate::json::JsonOf<StreamPatch>>,
    },
}

impl StreamStreamCommands {
    pub async fn exec(
        self,
        client: &Svix,
        color_mode: colored_json::ColorMode,
    ) -> anyhow::Result<()> {
        match self {
            Self::List { options } => {
                let resp = client.stream().stream().list(Some(options.into())).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Create { stream_in, options } => {
                let resp = client
                    .stream()
                    .stream()
                    .create(stream_in.into_inner(), Some(options.into()))
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Get { stream_id } => {
                let resp = client.stream().stream().get(stream_id).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Update {
                stream_id,
                stream_in,
            } => {
                let resp = client
                    .stream()
                    .stream()
                    .update(stream_id, stream_in.into_inner())
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Delete { stream_id } => {
                client.stream().stream().delete(stream_id).await?;
            }
            Self::Patch {
                stream_id,
                stream_patch,
            } => {
                let resp = client
                    .stream()
                    .stream()
                    .patch(stream_id, stream_patch.unwrap_or_default().into_inner())
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
        }

        Ok(())
    }
}
